### ISSUE TITLE
Index mask

### ISSUE THREAD

**Issue by Durman** [nortikin/sverchok#2764]
## Problem statement

According this conversation #2649, Sverchok is very weak in manipulating with different parts of one mesh. At this moment we split such parts to different meshes for manipulate them freely (moving, scaling or whatever). Such approach has big downside that it can't be vectorized and this process is unlike how users used to deal in Blender.

## Solution

Solution I think can be very easy. The goal is to mark different parts of mesh somehow so they can be differed from each other. Most simple way is creating index mask. It can be vertex index mask, edges index mask but for beginning it is better have face index mask above all. 
Face index mask will be like mask which we already have but instead of True or False values it will be have integer values from 0 to infinity. Index will be mean to which part of mesh current face is related.

## Example

Imagine we have added box generator which generates boxes in one object. With object it will generate face index mask. For first box it will generate mask like this `[0, 0, 0, 0, 0, 0]` where 0 means flag of part of mesh and number of values is equal to number of generated face. For second box will be generated the same mask but with next index 1. And the same for all other boxes. For 3 boxes faces mask index output will look in this way: `[0,0,0,0,0,0,1,1,1,1,1,1,2,2,2,2,2,2]`.

Such mask already gives opportunity to manipulate parts of mesh with existing tools of Sverchok. Firstly, it is possible to **converts index mask to normal mask** which will present selection of one or several parts of mesh. In this example boxes was moved by `transform selected` node:

![2019-12-24_09-05-46](https://user-images.githubusercontent.com/28003269/71394932-2edf6e80-262d-11ea-9545-a10615cdbd21.png)

Another opportunity is to **distribute values**. For example we would like to insert faces to the mesh but first part should not have extra faces at all and second and third parts of meshes should have different value of offsetting.
It is possible to do with `list item` node. In data socket values should plugged. In index socket should be plugged face index mask.

![2019-12-24_09-19-33](https://user-images.githubusercontent.com/28003269/71395268-bed1e800-262e-11ea-9b2a-7cfdb0812a6b.png)

This two approach gives total control over parts of mesh. For opportunity manipulate parts of objects in space freely `apply matrix` nodes should have modes apply matrix per vertex, per edge and per face and `matrix` node should be vectorized as it already is.

**Comment by portnov** [comment]
I do not see, why box generator should/would generate three boxes in one object?

**Comment by Durman** [comment]
It should not. It should generates boxes however we want. It should have opportunity ot generate boxes in different object and in one object as well. Benefits of having opportunity to generate primitives in one object I have described above.

**Comment by portnov** [comment]
You can generate that "index mask" together with multiple objects as well: `[[0,0,0], [1,1,1]...]`.
Imho if you wish to have several boxes in one objects, you can always use "mesh join" node.

**Comment by Durman** [comment]
If `mehs join` is used in Sverchok layout anywhere this layout does not supports vectorization anymore. For example it is impossible to generate first object with one box second object with two boxes and third objects with three boxes with `mesh join` node.

**Comment by zeffii** [comment]
I don't think this should be default behavior for generator nodes, but we could definitely accommodate better "sub object tracking".

to help us better understand, i propose that you make a `mesh join index out` node, which can merge f.ex 3 objects into one (like `verts.3` becomes `verts.1`) and it outputs `vert_obj_indices.1` (a flat list that tells us which object each element is associated with), same for edges / faces.

```text

  |__mesh join indices__|
  +---------------------+
  |            verts. 1 |
  |            edges. 1 |
  |            faces. 1 |
  |            v_idx. 1 |
  |            e_idx. 1 |
  |            f_idx. 1 |
  |                     |
  | verts. 3            |
  | edges. 3            |
  | faces. 3            |
  | matrices.           |
```

**Comment by zeffii** [comment]
but a more compressed format would be 
```
[0, 0, 0, 0, 0, 0, 0, 0 ,1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2]
=
[[8, 0], [10, 1], [7, 2]]
=
[8, 10, 7]
```

**Comment by zeffii** [comment]
i'm entertaining this, because the notion / notation has a way of re-emerging every few years. maybe for a good reason.

it might be easier to warm to the idea if I consider an additional output socket for Generators, a 'decoder' socket. which would hold a dict of such indices, so that you can output multiple sub_objects in one vertex list ( ` verts. 1 `)
```
decoder_dict = dict(
    verts=[.........],
    edges=[........]
    faces=[.....]
)
self.outputs['decoder'].sv_set(decoder_dict)

```
but, with the `encoder` node that can split them up if needed, based purely on the `received dict` from the decoder socket, and not purely topologically separated.

```
ability to filter, and invert filer

```

**Comment by Durman** [comment]
I don't think for developing current concept it is worth to start from adding index mask sockets or something to generator nodes. We have another nodes which can generates such indexes and logic of generation such indexes for generators is quite simple. I would rather to think where it is possible to apply such indexes mask and it looks like that the node which could move differen parts of a mesh would be good start.

```text

  |__transform partly__|
  +--------------------+
  |            verts. |
  |            edges. |
  |            faces. |
  |                   |
  |  Index mask:      |
  |  vert/edge/face   |
  |                   |
  | verts.            |
  | edges.            |
  | faces.            |
  | index mask        |
  | matrices.         |
```
About format: I don't think it can be simplified. Mask like this [0,0,0,1,1,1,2,2,2] is corner case and can have a look like [0,2,0,1,1,2,0,2,1] after some topological changes for example or any other cases where order of verteces/edges/faces is changing.

Adding such information about indexes can stretch nodes too wide. And it is not only about this proposal. We already have started to add face data information to nodes, probably edges and vertices data will be added soon to some nodes. It is also make nodes too long. I sink for this problem it is worth to open another issue. (And vert/edge/face index mask can be part of vert/edge/face data.)

**Comment by Durman** [comment]
Using index mask for assigning material to separate parts of mesh:
![2020-01-13_13-11-26](https://user-images.githubusercontent.com/28003269/72244444-b6564700-3607-11ea-8662-51727aaddfb4.png)

**Comment by Durman** [comment]
Before starting experiments with index mask I'm going create `Transform mesh` node which will work with regular selection masks. I will try to reimplement Blender transform operators as close as possible. Then probably some logic from Sverchok can be added.
![2020-01-13_16-37-12](https://user-images.githubusercontent.com/28003269/72256579-faa31080-3622-11ea-9c67-e184414ad748.png)
I don't think I'm going to use matrixes for simplicity. Matrixes are not vectorized as well.

**Comment by Durman** [comment]
flatten monkey by nearby point:
![2020-01-27_15-01-57](https://user-images.githubusercontent.com/28003269/73169759-3905f880-4116-11ea-9c8d-d565371ff7a9.png)

**Comment by Durman** [comment]
Now we have `separate parts to indexes` node which can generate index masks and `transform mesh` node which can transform mesh according given index mask.
